File: 0900-0999/981-time-based-key-value-store.cpp
class TimeMap {
public:
    TimeMap() {
        
    }
    vector<pair< string, pair<string, int>>> v;
    
    void set(string key, string value, int timestamp) {
        pair<string, int> p = {value, timestamp};
        pair<string, pair<string, int> > p1 = {key, p};
        v.push_back(p1);
   
    }
    
    string get(string key, int timestamp) {
        for(int i=v.size()-1; i>=0; i--) {
            if(v[i].first==key) {
                if(v[i].second.second==timestamp) {
                    return v[i].second.first;
                }
                else if(v[i].second.second<timestamp) {
                    return v[i].second.first;
                }
            }
        }
        return "";
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */

File: 0900-0999/981-time-based-key-value-store.py
# https://leetcode.com/problems/time-based-key-value-store/description/

class TimeMap:

    def __init__(self):
        self.dict = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.dict[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:
        values = self.dict.get(key, [])
        if not values:
            return ""

        left = 0
        right = len(values) - 1

        while left < right:
            mid = left + (right - left) // 2
        
            if values[mid][1] > timestamp:
                right = mid
            else:
                left = mid + 1

        if values[left][1] > timestamp:
            left -= 1

        return values[left][0] if values[left][1] <= timestamp else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
